### PR TITLE
Improvement: Clicked blocks highlight

### PR DESCRIPTION
**Pull request by Vahvl**
## What
Added a lot of functionality to the previously bland highlight clicked blocks feature. All colors are now changeable using moulconfig's color thingy and the text part can be toggled without removing the highlights. There is now a chat message based detection for when a chest is locked and a seperate color can be picked for that, it has a very small delay between the normal chest highlight showing up and it changing to the locked one. Kept the random aspect of the original behind another toggle and it now has a lot more colors it can choose. NEU chroma doesn't quite work, acts as more of a color randomizer.

A way to have the lever and wither essence highlight be smaller would be cool but there doesn't seem to be a simple way of doing that.

tldr: very more goog

## Changelog Improvements
+ Improved the clicked block highlight feature. - Vahvl & j10a1n15
    * Added extensive customization options for colors, locked chest detection, and more.

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/config/features/dungeon/HighlightClickedBlocksConfig.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    @ConfigOption(name = "Show Text", desc = "If enabled shows a text telling you what you clicked with the highlight.")
+    @ConfigEditorBoolean
+    public boolean showTextEnabled = true;
+
+    @Expose
+    @ConfigOption(name = "Random Colour", desc = "If enabled makes the colours random.")
+    @ConfigEditorBoolean
+    public boolean randomColourEnabled = false;
+
+    @ConfigOption(name = "Reset Colors", desc = "Resets the colors of the highlights to default ones.")
```
Comment (ItsEmpa) [inline src/main/java/at/hannibal2/skyhanni/config/features/dungeon/HighlightClickedBlocksConfig.java:52]:
```suggestion
    @ConfigOption(name = "Reset Colours", desc = "Resets the colours of the highlights to default ones.")
```

File: src/main/java/at/hannibal2/skyhanni/features/dungeon/DungeonHighlightClickedBlocks.kt [hunk 2]
```diff
@@ -1,56 +1,73 @@
...
-        if (!DungeonAPI.inDungeon()) return
+        if (!isEnabled()) return
 
         if (event.message == "§cYou hear the sound of something opening...") {
             event.blockedReason = "dungeon_highlight_clicked_block"
         }
+
+        if (event.message == "§cThat chest is locked!") {
+            blocks.lastOrNull { it.value.displayText.contains("Chest") }?.value?.colour = config.lockedChestColour.toChromaColor()
+        }
```
Comment (jani270) [inline src/main/java/at/hannibal2/skyhanni/features/dungeon/DungeonHighlightClickedBlocks.kt]: These should prob get converted to repo patterns

File: src/main/java/at/hannibal2/skyhanni/features/dungeon/DungeonHighlightClickedBlocks.kt [hunk 3]
```diff
@@ -69,28 +86,38 @@ object DungeonHighlightClickedBlocks {
...
-        CHEST("Chest"),
-        WITHER_ESSENCE("Wither Essence"),
+    @SubscribeEvent
+    fun onConfigFix(event: ConfigUpdaterMigrator.ConfigFixEvent) {
+        event.move(52, "dungeon.highlightClickedBlocks", "dungeon.clickedBlocks.enabled")
     }
+
+    class ClickedBlock(val displayText: String, var colour: Color)
+
+    fun isEnabled() = DungeonAPI.inDungeon() && config.enabled
```
Comment (ItsEmpa) [inline src/main/java/at/hannibal2/skyhanni/features/dungeon/DungeonHighlightClickedBlocks.kt]:
```suggestion
    fun isEnabled() = !DungeonAPI.inBossRoom && inDungeonAPI.inDungeon() && config.enabled

```

File: src/main/java/at/hannibal2/skyhanni/features/dungeon/DungeonHighlightClickedBlocks.kt [hunk 4]
```diff
@@ -1,56 +1,73 @@
...
+            blocks.lastOrNull { it.value.displayText.contains("Chest") }?.value?.colour = config.lockedChestColour.toChromaColor()
+        }
     }
 
     @SubscribeEvent
     fun onBlockClick(event: BlockClickEvent) {
-        if (!SkyHanniMod.feature.dungeon.highlightClickedBlocks) return
-        if (!DungeonAPI.inDungeon()) return
+        if (!isEnabled()) return
         if (DungeonAPI.inBossRoom) return
```
Comment (ItsEmpa) [inline src/main/java/at/hannibal2/skyhanni/features/dungeon/DungeonHighlightClickedBlocks.kt]:
```suggestion
```

**Review by ItsEmpa — CHANGES_REQUESTED** [review]
some nitpicks

File: src/main/java/at/hannibal2/skyhanni/config/features/dungeon/HighlightClickedBlocksConfig.java (hunk 1 above)
Comment (jani270) [reply src/main/java/at/hannibal2/skyhanni/config/features/dungeon/HighlightClickedBlocksConfig.java:52]: I Disagree

Comment (ItsEmpa) [reply src/main/java/at/hannibal2/skyhanni/config/features/dungeon/HighlightClickedBlocksConfig.java:52]: i thought we said for now we would use colour when its user facing

Comment (jani270) [reply src/main/java/at/hannibal2/skyhanni/config/features/dungeon/HighlightClickedBlocksConfig.java:52]: /sh color vs /sh colour disagrees with that

Comment (ItsEmpa) [reply src/main/java/at/hannibal2/skyhanni/config/features/dungeon/HighlightClickedBlocksConfig.java:52]: well all other configs in this class say colour instead of color, so either this one changes to colour or all the other ones change to color

File: src/main/java/at/hannibal2/skyhanni/features/dungeon/DungeonHighlightClickedBlocks.kt (hunk 2 above)
Comment (Vahvl) [reply src/main/java/at/hannibal2/skyhanni/features/dungeon/DungeonHighlightClickedBlocks.kt]: how would i do that?

Comment (jani270) [reply src/main/java/at/hannibal2/skyhanni/features/dungeon/DungeonHighlightClickedBlocks.kt]: > how would i do that?

Shift + CTRL + F for RepoPattern.pattern in SkyHanni code base and look at any off them

**Review by ItsEmpa — APPROVED** [review]

**Comment by Vahvl** [comment]
hanni put me back in 0.27 pls ive been good

File: src/main/java/at/hannibal2/skyhanni/features/dungeon/DungeonHighlightClickedBlocks.kt [hunk 5]
```diff
@@ -69,28 +98,37 @@ object DungeonHighlightClickedBlocks {
...
 
-    class ClickedBlock(val position: LorenzVec, val displayText: String, val color: LorenzColor, val time: Long)
-
-    enum class ClickedBlockType(val display: String) {
-        LEVER("Lever"),
-        CHEST("Chest"),
-        WITHER_ESSENCE("Wither Essence"),
+    @SubscribeEvent
+    fun onConfigFix(event: ConfigUpdaterMigrator.ConfigFixEvent) {
+        event.move(52, "dungeon.highlightClickedBlocks", "dungeon.clickedBlocks.enabled")
```
Comment (j10a1n15) [inline src/main/java/at/hannibal2/skyhanni/features/dungeon/DungeonHighlightClickedBlocks.kt:127]: the config fix event is outdated
